[libc][POSIX] Add clock_settime() function for Linux

amemov opened the pull request:
Closes #161461
- This is my first time contributing to libc's POSIX, so for reference I used `clock_gettime` implementation for Linux. For convenience, here is the description of `clock_settime` function [behavior](https://www.man7.org/linux/man-pages/man3/clock_settime.3.html)

amemov committed:
- Implemented a clock_settime syscall wrapper for Linux
- Implemented unit-tests for the feature

amemov commented:
Here are some questions that popped up when I started working on my PR:
- When I was looking at the `clock_gettime` implementation, I saw that it is implemented for GPUs, but in [`time.rst`](https://github.com/llvm/llvm-project/blob/4e404d0e5135ae7d0d4e7cc4c9b04d456eb50a01/libc/docs/headers/time.rst) file it is not marked as check. Is it correct or I should change that in my PR?
- About testing: from my understanding to call `clock_settime()` and actually change the time succesfully, a caller needs to have appropriate permission. How to emulate this in tests or it can be ignored?

amemov committed:
- Fixed CMake to include the implementation only on Linux for now

amemov commented:
@lntue

lntue commented on libc/src/__support/time/CMakeLists.txt:
```diff
@@ -19,3 +19,12 @@ add_object_library(
     libc.src.__support.time.${LIBC_TARGET_OS}.clock_gettime
 )
+
+if(LIBC_TARGET_OS_IS_LINUX)
```
change to
```
if(TARGET libc.src.__support.time.${LIBC_TARGET_OS}.clock_settime)
```

amemov commented on libc/src/time/linux/clock_settime.cpp:
```diff
@@ -0,0 +1,30 @@
+namespace LIBC_NAMESPACE_DECL {
+
+LLVM_LIBC_FUNCTION(int, clock_settime,
+                   (clockid_t clockid, const struct timespec *ts)) {
```
I mean technically you're right - I wrote it like that to be consistent with other implementation files in that folder, since all of them in parameters say `struct` keyword

lntue replied:
All of them will need to be removed, unless there are struct-function-variable name conflicts.

amemov replied:
Done - I also fixed it for other implementations in the folder, so everything stays consistent.

amemov committed:
- Changed CMake for clock_settime() to be more flexible for future targets

amemov committed:
- Removed redundant 'struct' keyword

amemov committed:
clang-format

michaelrj-google commented on libc/test/src/time/clock_settime_test.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "test/UnitTest/Test.h"
+
+using LlvmLibcClockSetTime = LIBC_NAMESPACE::testing::ErrnoCheckingTest;
+
```
there should also be a test where the syscall succeeds, we want to check both sides of the branch.

michaelrj-google reviewed (commented):
Overall looks good

amemov commented on libc/src/__support/time/linux/clock_settime.cpp:
```diff
@@ -0,0 +1,53 @@
+      sizeof(time_t) == sizeof(int64_t),
+      "SYS_clock_settime64 requires struct timespec with 64-bit members.");
+
+  __kernel_timespec ts64{};
+
+  ret = LIBC_NAMESPACE::syscall_impl<int>(SYS_clock_settime64,
+                                          static_cast<long>(clockid),
+                                          reinterpret_cast<long>(&ts64));
+  if (ret == 0) {
+    ts->tv_sec = static_cast<decltype(ts->tv_sec)>(ts64.tv_sec);
+    ts->tv_nsec = static_cast<decltype(ts->tv_nsec)>(ts64.tv_nsec);
+  }
```
@SchrodingerZhu , what do you mean?

amemov committed:
- Added positive path for clock_settime test
- Fixed error in clock_settime struct setup
- Upd the time.rst doc with up to date info

SchrodingerZhu reviewed (approved):
LGTM. Sorry for the delayed reply.

michaelrj-google reviewed (approved):
LGTM, do you want me to merge this for you?
